backend/hct_mis_api/apps/program/api/serializers.py

```diff
@@ -0,0 +1,166 @@
+
+    @staticmethod
+    def get_program(program_id: str) -> Program:
+        program = get_object_or_404(Program, id=decode_id_string(program_id))
+        return program
+
+    def validate_title(self, value: str) -> str:
+        program = self.get_program(self.context["request"].parser_context["kwargs"]["program_id"])
+        cycles = program.cycles.all()
+        if cycles.filter(title=value).exists():
+            raise serializers.ValidationError({"title": "Programme Cycles' title should be unique."})
+        return value
+
+    def validate(self, data: Dict[str, Any]) -> Dict[str, Any]:
+        program = self.get_program(self.context["request"].parser_context["kwargs"]["program_id"])
+        data["program"] = program
+        data["created_by"] = self.context["request"].user
+        start_date = data.get("start_date")
+        end_date = data.get("end_date")
+
+        if program.status != Program.ACTIVE:
+            raise serializers.ValidationError("Create Programme Cycle is possible only for Active Programme.")
+
+        if start_date and start_date < program.start_date:
+            raise serializers.ValidationError(
+                {"start_date": "Programme Cycle start date cannot be earlier than programme start date"}
+            )
+        if end_date and end_date > program.end_date:
+            raise serializers.ValidationError(
+                {"end_date": "Programme Cycle end date cannot be later than programme end date"}
+            )
+
+        if program.cycles.filter(end_date__isnull=True).exists():
+            raise serializers.ValidationError("All Programme Cycles should have end date for creation new one.")
+
+        # timeframes overlapping
+        validate_cycle_timeframes_overlapping(program, start_date, end_date)
+        return data
+
+
+class ProgramCycleUpdateSerializer(EncodedIdSerializerMixin):
+    title = serializers.CharField(required=False)
+    start_date = serializers.DateField(required=False)
+    end_date = serializers.DateField(required=False)
+
+    class Meta:
+        model = ProgramCycle
+        fields = ["title", "start_date", "end_date"]
+
+    def validate_title(self, value: str) -> str:
+        value = value.strip()
+        if (
+            ProgramCycle.objects.filter(title=value, program=self.instance.program, is_removed=False)
+            .exclude(id=self.instance.id)
+            .exists()
+        ):
+            raise serializers.ValidationError("A ProgramCycle with this title already exists.")
+
+        if value is None or "":
+            raise serializers.ValidationError("Not possible leave the Programme Cycle title empty.")
+
+        return value
+
+    def validate(self, data: Dict[str, Any]) -> Dict[str, Any]:
+        program = self.instance.program
+        if program.status != Program.ACTIVE:
+            raise serializers.ValidationError("Update Programme Cycle is possible only for Active Programme.")
+
+        if data.get("start_date") is None:
```

Comment:
remove this one as well because it's not possible to set None because `allow_null=False` is by default.
User allowed update all three fields and on edit level all of this are not required.